haskoin-wallet/Network/Haskoin/Wallet/Accounts.hs

```diff
@@ -124,56 +126,63 @@ initGap accE = do
     unless (validAccountType newAccountType) $
         throwM $ WalletException "Invalid account type"
-    let gen = isNothing newAccountMnemonic &&
-              isNothing newAccountMaster &&
-              null newAccountKeys
-    (mnemonicM, masterM, keys) <- if gen
-        then do
-            when (isJust newAccountMaster || isJust newAccountMnemonic) $
-                throwM $ WalletException
-                "Master key or mnemonic not allowed for generate"
-            ent <- liftIO $ getEntropy 16
-            let ms = fromRight $ toMnemonic ent
-                root = makeXPrvKey $ fromRight $ mnemonicToSeed "" ms
-                master = case newAccountDeriv of
-                    Nothing -> root
-                    Just d  -> derivePath d root
-                keys = deriveXPubKey master : newAccountKeys
-            return (Just ms, Just master, keys)
-        else case newAccountMnemonic of
-             Just ms -> do
-                 when (isJust newAccountMaster) $ throwM $ WalletException
-                     "Cannot provide both master key and mnemonic"
-                 root <- case mnemonicToSeed "" (cs ms) of
-                     Right s -> return $ makeXPrvKey s
-                     Left _ -> throwM $ WalletException
-                         "Mnemonic sentence invalid"
-                 let master = case newAccountDeriv of
-                         Nothing -> root
-                         Just d -> derivePath d root
-                     keys = deriveXPubKey master : newAccountKeys
-                 return (Nothing, Just master, keys)
-             Nothing -> case newAccountMaster of
-                 Just master -> do
-                     let keys = deriveXPubKey master : newAccountKeys
-                     return (Nothing, newAccountMaster, keys)
-                 Nothing -> return (Nothing, newAccountMaster, newAccountKeys)
+
+    when (isJust newAccountMnemonic && isJust newAccountMaster) $ throwM $
+        WalletException "Can not set both master key and mnemonic"
+
+    (mnemonicM, rootM) <- case newAccountMaster of
+        Just _ -> return (Nothing, newAccountMaster)
+        _ -> case newAccountMnemonic of
+            Just ms -> do
+                root <- mnemonicToPrvKey $ cs ms
+                -- Don't return our own mnemonic as we don't want it displayed
+                return (Nothing, Just root)
+            _ -> if null newAccountKeys
+                    then do
+                        -- Generate a key if nothing was provided
+                        let defEnt = configEntropy def
+                        ms   <- genMnemonic $ fromMaybe defEnt newAccountEntropy
+                        root <- mnemonicToPrvKey ms
+                        return (Just ms, Just root)
+                    else return (Nothing, Nothing)
+
+    let masterM = case newAccountDeriv of
+                      Nothing -> rootM
+                      Just d  -> (derivePath d) <$> rootM
```

Comment:
Derivations are computer server side here if a path was provided in the API call. I think this makes the API a bit more general.